### PR TITLE
openshift-install-manifest invoker set with ARO

### PR DESCRIPTION
**Pull request by gvanderpotte**
### Which issue this PR addresses:

Implements changes to cover https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9376061/

### What this PR does / why we need it:

This PR adds the ARO operator with the ability to reconcile the openshift-install-manifest ConfigMap in openshift-config namespace and set the invoker key to ARO. This value will hence be pushed by the CVO as a metric which can be read by Telemetry to track what created the cluster.

### Test plan for issue:

- Are there unit tests? yes
- Are there integration/e2e tests? Not done
- If it is not possible to write automated tests, explain why and document how

+ spin up new cluster
+ check openshift-install-manifests ConfigMap invoker value
+ start local operator
+ check ConfigMap once again
+ invoker should be set to ARO
+ Get the "cluster_install" metric from openshift-k8s, similar to what is done in https://bugzilla.redhat.com/show_bug.cgi?id=1802331
  oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" 'https://prometheus-k8s.openshift-monitoring.svc:9091/api/v1/query?query=cluster_installer' | jq
   > cluster_install metric has the invoker set to "ARO"

### Is there any documentation that needs to be updated for this PR?

Probably not

### CODE REVIEW

File: pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+func (r *OpenshiftInstallReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	return reconcile.Result{}, r.setOpenshiftInstallInvoker(ctx, "AR0")
+}
+
+// setOpenshiftInstallInvoker sets data.invoker for openshift-install-manifest to invoker string. This will be picked up by cluster-version-operator to generate cluster_installer metric
+// and then be taken by Telemetry
+func (r *OpenshiftInstallReconciler) setOpenshiftInstallInvoker(ctx context.Context, invoker string) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		oim, err := r.kubernetescli.CoreV1().ConfigMaps(openshiftInstallConfigMap.Namespace).Get(ctx, openshiftInstallConfigMap.Name, metav1.GetOptions{})
```
Comment (petrkotas) [inline pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go:49]: This get can get out of retry, gets are not retry able.

File: pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	return &OpenshiftInstallReconciler{
+		kubernetescli: kubernetescli,
+		log:           log,
+	}
+}
+
+// Reconcile makes sure that the openshift-install-manifest ConfigMap has the right value for invoker key.
+func (r *OpenshiftInstallReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	return reconcile.Result{}, r.setOpenshiftInstallInvoker(ctx, "AR0")
```
Comment (petrkotas) [inline pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go:42]: Can "ARO" be a `const` defined on top near the `var openshiftInstallConfigMap`.

File: pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		if inv, ok := oim.Data["invoker"]; ok {
+			if inv == invoker {
+				// Invoker already has the right value, no need for update
+				return nil
+			}
+		}
+
+		oim.Data["invoker"] = invoker
+
+		_, err = r.kubernetescli.CoreV1().ConfigMaps(openshiftInstallConfigMap.Namespace).Update(ctx, oim, metav1.UpdateOptions{})
```
Comment (petrkotas) [inline pkg/operator/controllers/openshiftinstall/openshiftinstall_controller.go:63]: If this succeeds it updates the object, but it does not leave trace for us to see it, only when we now where to look. 

As this alters the state of the cluster, I would propose adding a condition to the operator that signals this change.

**Review by petrkotas — APPROVED** [review]
This is really clever solution to the problem. The metric will pick the change and reflect it. Will the previous state be also significantly present?

I would love to see some sort of embedded documentation as part of the PR. 
Explanation at the top of the go file would be perfect. This way I would not have to think twice why this exists.

**Review by petrkotas — CHANGES_REQUESTED** [review]
Sorry, I clicked wrong button, please look at my previous comments.

**Comment by jim-minter** [comment]
@gvanderpotte sorry, a requirement got dropped on the floor here :-(  It is possible that now or in the future clusters will reach out to OCM *before* the ARO operator runs, e.g. to send installer telemetry.  We would like to ensure that the cluster type as seen by OCM does not change, so this configuration needs to happen earlier in the process.

Grep the installer source code for `os.Getenv("OPENSHIFT_INSTALL_INVOKER")`.  Naively, we would just do `os.Setenv("OPENSHIFT_INSTALL_INVOKER", "ARO")` in our code before calling the installer, but it's a bit clunky.  I would rather carry a small patch in our installer branch which sets "ARO" in the installer graph generation.  That should handle new cluster creations.  I'm not sure how much I care about updating the configs for existing clusters, but if you do I'd be tempted to do it as a one-shot admin upgrade step rather than worrying about the overhead of a controller for the rest of time.

While you're in the installer, note that we also write the string "was not built correctly" in the config.  I don't know if we want to do something about that.  It looks ugly, beyond that I don't care too much.  See if there's any lightweight worthwhile fix to be done there?

**Comment by gvanderpotte** [comment]
Thanks Jim for the comment. This requirement makes it impossible to use operator to handle that. I will do with installer patch then.

**Pull request closed without merging**